### DISCUSSION TITLE
Mark image for pruning

### DISCUSSION THREAD

**Discussion by ygalblum** [containers/podman#18182] [Ideas]
I have a use case in which I would like to pull a new version of an image, but not use it until reboot (or an explicit service restart). As a result, since the image may be currently in use, I cannot remove the old image during the same process.

Podman offers an `images prune` command that may be used to remove dangling images or even all unused images (by passing `--all`). But, removing all unused images is not the right solution in an environment with limited connectivity as it might remove images that will be used in the future, just not at the time `prune` is called. 

Alternatively, the `prune` command also supports a `filter` argument to prune only the images containing certain labels (or label=value). However, labels are set during the creation of the image and cannot be altered. So, this path is blocked.

So, I would like to add a command that will mark images for removal and then an option to prune to remove such images.

**Comment by Luap99** [comment]
If you know what image you want to remove, is there any reason to not just call `podman rmi` on it?

**Comment by ygalblum** [comment]
The reason I can't call `podman rmi` is that the call will fail if the image is in use at the time that I'm pulling the new version.

I'll give an example:

Let's assume my Quadlet file is:
```
[Container]
Image=ygal:0.1
```
So, podman will be running the `ygal:0.1` image.

Now, let's say I would like to upgrade by pulling the new image `ygal:0.2` and updating my Quadlet file accordingly:
```
[Container]
Image=ygal:0.2
```
However, I do not wish for the upgrade to happen at the time that I'm pulling the update, but only after reboot (or after the user explicitly ask for it). Please keep in mind that changes to the Quadlet file are not processes by Quadlet until either a reboot or a call to `systemctl daemon-reload`.

So, when my update phase is complete, I have both images locally, `ygal:0.1` is in use and `ygal:0.2` is ready to be used after the reboot. As a result, during the update phase, calling `podman rmi` will fail. Once a reboot happens, Quadlet will process the updated file and `ygal:0.2` will be used. `ygal:0.1` will stay there without anyone knowing it should be removed.

So, what I am suggesting here is the ability to mark `ygal:0.1` for pruning and the ability for `images prune` to remove only the images marked as such.

I think the alternative will be to implement this mechanism separately. That is, maintain a list of images marked for pruning and have a process run `podman rmi` on them. But, I think having this ability baked into `podman` is more natural

**Reply by vrothberg** [reply]
I wonder how this "marking" should work.  `image prune` will only remove dangling images which are those without a name/tag and not used by any container.  So the "marking" could happen by "image untag".

But who takes care of that?

**Reply by vrothberg** [reply]
It could be made part of `image pull` where Podman would untag (or even remove) images with the same name but a different tag.

**Reply by ygalblum** [reply]
The problem with un-tagging is that it might break the ability to manage the currently running service.

I'll continue my example:

Let's say my Quadlet file's name is `ygal.container`. Quadlet will create a systemd unit file called `ygal.service` that the user can start/stop/restart. The service file will indicate `ygal:0.1` as the image to run in the `podman run` command set for `ExecStart`.

As explained before, following the upgrade, `ygal.service` will be regenerated only after reboot (or when `systemctl daemon-reload`) is called. As a result, until that happens, calling `systemctl restart ygal.service` will try to use `ygal:0.1`. Therefore, it cannot be removed or un-tagged

**Reply by vrothberg** [reply]
Who would take care of "marking" the image as "unused" in this scenario?

So we're strictly speaking about changes to a Quadlet-managed file.  Because I also see the use-case of `ygal:0.1` being updated from `0.1.0` to `0.1.123` and the old images still lurking around.  It's common for container tags to have multiple tags for the same image, for instance, `name:1.2.3` also being tagged as `name:1.2` and `name:1` (depending on the time of update).

**Reply by ygalblum** [reply]
The same process that will call `podman pull` to get the new image is the one that will mark the old image.

I don't know if this is Quadlet only case. It's just the scenario I had in mind because when using Quadlet, the user does not need to run anything other than reboot in order for the changes to take affect. In other cases, I assume the user will have to run something to use the new container and then she can also remove the image.

As for the different tags for the same actual image, not sure what you mean.
Let's say we have `name:1` and `name:1.2` pointing to the same image. So, the update process will pull `name:1.2` (which will result in not downloading anything) and marking `name:1` for removal. So, when `podman prune` removes `name:1`, won't it result in only removing the `name:1` tag but not actually removing `name:1.2`. What is the current behavior for `prune --all` in such a case, assuming `name:1` is not used while `name:1.2` is.

**Reply by ygalblum** [reply]
@vrothberg sorry, I kinda lost track on this one. Can we restart the discussion?

**Reply by vrothberg** [reply]
> Let's say we have name:1 and name:1.2 pointing to the same image. So, the update process will pull name:1.2 (which will result in not downloading anything) and marking name:1 for removal. So, when podman prune removes name:1, won't it result in only removing the name:1 tag but not actually removing name:1.2. What is the current behavior for prune --all in such a case, assuming name:1 is not used while name:1.2 is.

If an image has multiple tags and only one is used by a container, it won't get pruned and nothing gets untaged.

I sent you an invite to chat about this idea next week.

**Reply by ygalblum** [reply]
Summing up the chat @vrothberg and I had.

Marking images for pruning is problematic. While one service may move to a new image, deeming the old one as pruneable, a different service might still use the old one. As a result, the updated service should not be able to mark the image as pruneable. 

This issues moved the discussion to using ref-counts. However, this seemed a bit too complex.

Instead we thought about the following solution:

Each service knows which images it needs and therefore should not be removed when pruning even if they are not currently running. Therefore, Podman should provide a way to protect images from being pruned.

The way to do so is by adding a configuration field in containers.conf in the `engine` section called `unpruneable_images`. Each service can then drop a uniquely named file under `/usr/share/containers/conf.d` listing the images it would like to block from pruning. For example:
```
[engine]
unpruneable_images = [
  "docker.io/library/wordpress:4.9-apache",
  "docker.io/envoyproxy/envoy:v1.25.0",
]
```

Podman will then process this field in an additive way. So, that its value will be the unique accumulation of all the instances of this field in all configuration files.

Then, Podman will not remove any image listed in `unpruneable_images` when `podman image prune --all` is called.

@rhatdan WDYT?

**Reply by Luap99** [reply]
> /usr/share/containers/conf.d

I don't think any application (besides package managers) should ever write under /usr/share. Also `/usr` is read only on all rpm-ostree based distros.

So the other option would be `/etc/containers/...`, but also based on past bugs I had to deal with there are systems were /etc is read only as well.

I like the idea overall but I think we need a different location for such things. Is there a reason to use containers.conf for this? I am worried about performance concerns if you consider that each podman command needs to parse all config files. So this adds overhead to all commands not just podman image prune.

**Reply by ygalblum** [reply]
@Luap99 thanks for the input.
@vrothberg and I were discussing this in the context of using RPMs to install the Quadlet files. That's why we though about using `/usr/share`. I am not sure, can the process of `rpm-ostree` write to `/usr`? If not, where should it go? Would `/var/lib/containers/` work?

As for performance, I see your point. Do you think having a separate configuration file used only by the `prune` command is a more viable solution?

**Reply by vrothberg** [reply]
I am not that concerned about the performance impact.  We are talking about a very controlled environment (see install via RPMs/ostree) and a narrow use case.  It's up to the users to decide whether they want to manage one big file or many small ones.

**Reply by vrothberg** [reply]
> Is there a reason to use containers.conf for this?

To me it seems like a natural candidate.  It's shared by buildah and podman and supports drop-in files etc.  Creating something similar from scratch seems unattractive and error prone. An alternative could be to use `.prune` files in `containers.conf.d`.  Config files must end with `.conf` so they would be ignored by default.  But as mentioned above, I am not worried about the performance impact.

**Reply by Luap99** [reply]
Yeah I am generally fine with a containers.conf approach. I am personally not too concerned about performance either, I just bring this up because automotive is.

Second, maybe I was missing the point that you want to install these files as rpm. I thought you want make quadlet create the files for each container unit?

---
Maybe a different direction, have you thought about making podman prune parse the quadlet files and ignore the images in there. This would make it work without any new extra files and all quadlet users could use it not just the files that are shipped via packaging.

**Reply by vrothberg** [reply]
> Second, maybe I was missing the point that you want to install these files as rpm. I thought you want make quadlet create the files for each container unit?

It's not necessarily tied to automotive where images would be part of the ostree image and pre-installed.  The use case is to ship containerized services as Quadlet files via RPMs.  I understand that's it's still WIP but the task was to find a way to prevent images from accidentally being pruned from the machine when an installed RPM is depending on it. Declaring that in a config file seemed like a straight forward and easy to support way of achieving that.

> Maybe a different direction, have you thought about making podman prune parse the quadlet files and ignore the images in there. This would make it work without any new extra files and all quadlet users could use it not just the files that are shipped via packaging.

We discussed the idea of `image prune` parsing the Quadlet (and Kube) files.  One downside is that it's easy to regress: new functionality in Quadlet may require extending the `prune` logic.

**Reply by Luap99** [reply]
> It's not necessarily tied to automotive where images would be part of the ostree image and pre-installed. The use case is to ship containerized services as Quadlet files via RPMs. I understand that's it's still WIP but the task was to find a way to prevent images from accidentally being pruned from the machine when an installed RPM is depending on it. Declaring that in a config file seemed like a straight forward and easy to support way of achieving that.

OK yeah this makes a sense then. Shipping them as rpm under /usr/share... should work just fine for this. With that knowledge I agree with the proposal to add a new containers.conf field and let the packagers/admins deal with adding the correct images there.

**Reply by vrothberg** [reply]
Since the need to mark images as unprunable came up recently, I want to highlight this conversation:
Cc: @rhatdan @baude @mheon @cgwalters

**Reply by cgwalters** [reply]
AFAICS this is (almost) exactly the same thing as https://github.com/containers/bootc/issues/128 

The (almost) is just around the detail of having multiple root filesystems in the bootc case, which doesn't exist in the package-based (apt/dnf) case.

**Comment by rhatdan** [comment]
Parsing the quadlet files would not be enough, you would also need to parse the kube.YAML files.  It would quickly get out of control. We might have to deal with additional stores as well.

I think potentially adding a last time used date to RW images could come in handy.  Might be useful for all users, what images are in my store that have not been used in one month, year, decade.
